pkg/binding/bindingpolicy-resolution.go

```diff
@@ -297,41 +267,43 @@ func workloadMatchesBindingSpec(bindingSpecWorkload *v1alpha1.DownsyncObjectRefe
 		}
 	}
 
-	if len(objectIdentifierToKeyMap) != clusterScopedObjectsCount+namespacedObjectsCount {
+	if len(objectIdentifiers) != clusterScopedObjectsCount+namespacedObjectsCount {
 		return false
 	}
 
 	// again we can check match by making sure all objects are mapped, since entries are unique and the length is equal.
 	// check cluster-scoped all exist
-	if !bindingClusterScopeIsMapped(bindingSpecWorkload.ClusterScope, objectIdentifierToKeyMap,
-		gvkGvrMapper) {
+	if !bindingClusterScopeIsMapped(bindingSpecWorkload.ClusterScope, objectIdentifiers) {
 		return false
 	}
 
 	// check namespace-scoped all exist
-	return namespaceScopeMatchesBindingSpec(bindingSpecWorkload.NamespaceScope, objectIdentifierToKeyMap,
-		gvkGvrMapper)
+	return namespaceScopeMatchesBindingSpec(bindingSpecWorkload.NamespaceScope, objectIdentifiers)
 }
 
 // bindingClusterScopeIsMapped returns true if the cluster-scope
 // section in the binding spec all exist in the resolution.
 func bindingClusterScopeIsMapped(bindingSpecClusterScope []v1alpha1.ClusterScopeDownsyncObjects,
-	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	objectIdentifiers sets.Set[util.ObjectIdentifier]) bool {
 	for _, clusterScopeDownsyncObjects := range bindingSpecClusterScope {
-		gvr := schema.GroupVersionResource(clusterScopeDownsyncObjects.GroupVersionResource)
-		gvk, found := gvkGvrMapper.GetGvk(gvr)
+		objectNamesSet := sets.New[string](clusterScopeDownsyncObjects.ObjectNames...)
 
-		if !found {
-			return false // if not found then not mapped and not in resolution
-		}
+		// TODO: optimize
+		for objIdentifier := range objectIdentifiers {
+			// skip objects that aren't cluster-scoped
+			if objIdentifier.ObjectName.Namespace != metav1.NamespaceNone {
+				continue
+			}
 
-		for _, objName := range clusterScopeDownsyncObjects.ObjectNames {
-			formattedKey := util.KeyFromGVKandNamespacedName(gvk, types.NamespacedName{
-				Namespace: metav1.NamespaceNone,
-				Name:      objName,
-			})
+			// skip objects that don't match the GVR
+			if objIdentifier.GVK.Group != clusterScopeDownsyncObjects.GroupVersionResource.Group ||
+				objIdentifier.GVK.Version != clusterScopeDownsyncObjects.GroupVersionResource.Version ||
+				objIdentifier.Resource != clusterScopeDownsyncObjects.GroupVersionResource.Resource {
+				continue
+			}
 
-			if _, found := objectIdentifierToKeyMap[formattedKey]; !found {
+			// check if object is in the objectNamesSet
+			if !objectNamesSet.Has(objIdentifier.ObjectName.Name) {
 				return false
```

Comment:
The logic in new lines 289 through here is backwards. These lines check whether the object names in clusterScopeDownsyncObjects are a _superset_ of those in objectIdentifiers with the right GVR. This func needs to check for subset rather than superset.

Reply:
Aren't the directions equivalent under the guarantee of uniqueness of entries in both groups (resolution's object identifiers, and namespaced + cluster scoped identifiers) and the equality of sizes?

The groups match if and only if every object in either group exists in the other, given equality of sizes.

Reply:
Ah, right.